# Regresión

### 1. **Predicción del número de heridos o muertos por día de la semana**  
   - **Variable objetivo**: `DIASEMANA`, `TOTHERIDOS`
   - **Variables predictoras**: `HORA`, `MES`, `TOTMUERTOS`, `TIPACCID`, `EDAD`, `CAUSAACCI`, `LONGITUD`, `LATITUD`  

En base a datos globales, se predecirá la cantidad de heridos por día de la semana (lunes, martes, etc.).

### 2. **Predicción de la probabilidad de heridos o muertos en un accidente**  
   - **Variable objetivo**: `PROBHERIDO`, `PROBMUERTO` Columnas que no existen y se harían a partir de feature engineering o a partir de la siguiente fórmula: 
   <br><br> $$P(\text{muerte}) = \frac{\text{TOTMUERTOS}}{\text{TOTMUERTOS} + \text{TOTHERIDOS}}$$  <br>
   - **Variables predictoras**: `TIPACCID`, `CLASE`, `CAUSAACCID`, `CAPAROD`, `ALIENTO`, `CINTURON`, `EDAD` 

# Clasificación

### 1. Causa de accidente
- **Variable objetivo:** `CAUSAACCI`
- **Variables predictoras:** `TOTMUERTOS`, `TOTHERIDOS`, `TIPOACCID`, `CLASE`, `Tipo de vehículo` (varias columnas), `MES`, `DIA`, `DIASEMANA`, `CAPAROD`, `ALIENTO`, `CINTURON`,
Predecir cuál fue la causa del accidente dando expliación a qué fue lo que ocasionó el accidente.

### 2. Tipo de accidente
- **Variable objetivo:** `TIPACCID`
- **Variables predictoras:** `TOTMUERTOS`, `TOTHERIDOS`, `CAUSAACCI`, `CLASE`, `Tipo de vehículo` (varias columnas), `MES`, `DIA`, `DIASEMANA`, `CAPAROD`, `ALIENTO`, `CINTURON`,
Definir el tipo de accidente que pudo ser causado por varias razones.

# Agrupamiento

### 1. Causa/Tipo de accidente
- **Variables principales:** `CAUSAACCID`, `TIPACCID`, `CLASE`, `LONGITUD`,`LATITUD`, etc.
Encontrar las principales causas y/o tipos de accidentes de acuerdo a las variables mencionadas.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('../data/raw/2018-2019/nacional/conjunto_de_datos/datos_mexico_2019.csv')

C:\Users\isaij\AppData\Local\Temp\ipykernel_36624\3372139521.py:1: DtypeWarning: Columns (0,46) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/raw/2018-2019/nacional/conjunto_de_datos/datos_mexico_2019.csv')


In [3]:
df2 = pd.read_csv('../data/raw/2018-2019/cdmx/hechos-de-transito-reportados-por-ssc-base-comparativa.csv')

# Datos nacionales

In [4]:
df.head()

ID  EDO  MES  ANIO  MPIO  HORA  MINUTOS  DIA  DIASEMANA  URBANA  ...  \
0   8    1    1  2019     1     4        0    3          4       1  ...   
1   9    1    1  2019     1     9       45    3          4       1  ...   
2  10    1    1  2019     1    10       40    3          4       1  ...   
3  11    1    1  2019     1    14       55    3          4       1  ...   
4  12    1    1  2019     1    19        0    3          4       1  ...   

   OTROMUERTO  OTROHERIDO  TOTMUERTOS  TOTHERIDOS  CLASE  \
0           0           0           0           0      3   
1           0           0           0           0      3   
2           0           0           0           0      3   
3           0           0           0           1      2   
4           0           0           0           0      3   

                                      CALLE1                        CALLE2  \
0  BOULEVARD LICENCIADO ADOLFO RUÍZ CORTINES           CALLE PASEO DEL RÍO   
1            AVENIDA PASEO DE LA SOLIDARIDAD  AVENIDA AGUASCALIENTES NORTE   
2                        AVENIDA UNIVERSIDAD                  CALLE ABEDUL   
3                     AVENIDA MAHATMA GANDHI        CALLE ABRAHAM GONZÁLEZ   
4             AVENIDA GLORIETA BENITO JUÁREZ       CALLE IGNACIO T. CHÁVEZ   

   CARRETERA   LONGITUD   LATITUD  
0        NaN -102.31962  21.87119  
1        NaN -102.28652  21.91522  
2        NaN -102.30621  21.90032  
3        NaN -102.29386  21.84853  
4        NaN -102.30055  21.87392  

[5 rows x 49 columns]

In [5]:
df.tail()

ID  EDO  MES  ANIO  MPIO  HORA  MINUTOS  DIA  DIASEMANA  URBANA  ...  \
180214  615   32   12  2019    56     1       39   25          3       1  ...   
180215  616   32   12  2019    56     2       50   27          5       1  ...   
180216  617   32   12  2019    56     9       37   27          5       1  ...   
180217  619   32   12  2019    56    18       40   30          1       1  ...   
180218  620   32   12  2019    56    15       45   31          2       1  ...   

        OTROMUERTO  OTROHERIDO  TOTMUERTOS  TOTHERIDOS  CLASE  \
180214           0           0           0           0      3   
180215           0           0           0           0      3   
180216           0           0           0           0      3   
180217           0           0           0           2      2   
180218           0           0           0           0      3   

                                                   CALLE1  \
180214                                    AVENIDA HIDALGO   
180215                        CALLE JESÚS GONZÁLEZ ORTEGA   
180216  CALZADA REYES HEROLES FRENTE FINCA URBANA NUME...   
180217                      CALZADA HÉROES DE CHAPULTEPEC   
180218               CALLE PASEO FRANCISCO GARCÍA SALINAS   

                                   CALLE2  CARRETERA   LONGITUD   LATITUD  
180214  CALLEJÓN CALLEJÓN DE GÓMEZ FARÍAS        NaN -102.57270  22.77540  
180215        CALLE VALENTÍN GÓMEZ FARÍAS        NaN -102.59222  22.76806  
180216                                NaN        NaN -102.57728  22.76134  
180217               CALLE MIGUEL CALDERA        NaN -102.59398  22.77347  
180218           CALLE PASEO DE LA CAÑADA        NaN -102.54566  22.75993  

[5 rows x 49 columns]

In [6]:
df.shape

(180219, 49)

In [7]:
df.columns

Index(['ID', 'EDO', 'MES', 'ANIO', 'MPIO', 'HORA', 'MINUTOS', 'DIA',
       'DIASEMANA', 'URBANA', 'SUBURBANA', 'TIPACCID', 'AUTOMOVIL', 'CAMPASAJ',
       'MICROBUS', 'PASCAMION', 'OMNIBUS', 'TRANVIA', 'CAMIONETA', 'CAMION',
       'TRACTOR', 'FERROCARRI', 'MOTOCICLET', 'BICICLETA', 'OTROVEHIC',
       'CAUSAACCI', 'CAPAROD', 'SEXO', 'ALIENTO', 'CINTURON', 'EDAD',
       'CONDMUERTO', 'CONDHERIDO', 'PASAMUERTO', 'PASAHERIDO', 'PEATMUERTO',
       'PEATHERIDO', 'CICLMUERTO', 'CICLHERIDO', 'OTROMUERTO', 'OTROHERIDO',
       'TOTMUERTOS', 'TOTHERIDOS', 'CLASE', 'CALLE1', 'CALLE2', 'CARRETERA',
       'LONGITUD', 'LATITUD'],
      dtype='object')

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180219 entries, 0 to 180218
Data columns (total 49 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   ID          180219 non-null  object 
 1   EDO         180219 non-null  int64  
 2   MES         180219 non-null  int64  
 3   ANIO        180219 non-null  int64  
 4   MPIO        180219 non-null  int64  
 5   HORA        180219 non-null  int64  
 6   MINUTOS     180219 non-null  int64  
 7   DIA         180219 non-null  int64  
 8   DIASEMANA   180219 non-null  int64  
 9   URBANA      180219 non-null  int64  
 10  SUBURBANA   180219 non-null  int64  
 11  TIPACCID    180219 non-null  int64  
 12  AUTOMOVIL   180219 non-null  int64  
 13  CAMPASAJ    180219 non-null  int64  
 14  MICROBUS    180219 non-null  int64  
 15  PASCAMION   180219 non-null  int64  
 16  OMNIBUS     180219 non-null  int64  
 17  TRANVIA     180219 non-null  int64  
 18  CAMIONETA   180219 non-null  int64  
 19  CA

In [9]:
df.describe()

EDO            MES      ANIO           MPIO           HORA  \
count  180219.000000  180219.000000  180219.0  180219.000000  180219.000000   
mean       16.373285       6.545558    2019.0      32.148148      12.750037   
std         7.100668       3.477758       0.0      30.395565       5.932928   
min         1.000000       1.000000    2019.0       1.000000       0.000000   
25%        11.000000       3.000000    2019.0      14.000000       8.000000   
50%        19.000000       7.000000    2019.0      26.000000      13.000000   
75%        19.000000      10.000000    2019.0      39.000000      18.000000   
max        32.000000      12.000000    2019.0     193.000000      23.000000   

             MINUTOS            DIA      DIASEMANA         URBANA  \
count  180219.000000  180219.000000  180219.000000  180219.000000   
mean       25.999723      15.630733       3.972334       1.019510   
std        17.756017       8.798444       1.966724       0.206587   
min         0.000000       1.000000       1.000000       0.000000   
25%        10.000000       8.000000       2.000000       1.000000   
50%        27.000000      15.000000       4.000000       1.000000   
75%        40.000000      23.000000       6.000000       1.000000   
max        59.000000      31.000000       7.000000       2.000000   

           SUBURBANA  ...     PEATHERIDO     CICLMUERTO     CICLHERIDO  \
count  180219.000000  ...  180219.000000  180219.000000  180219.000000   
mean        0.026157  ...       0.034147       0.000333       0.006242   
std         0.248610  ...       0.195754       0.018842       0.081054   
min         0.000000  ...       0.000000       0.000000       0.000000   
25%         0.000000  ...       0.000000       0.000000       0.000000   
50%         0.000000  ...       0.000000       0.000000       0.000000   
75%         0.000000  ...       0.000000       0.000000       0.000000   
max         3.000000  ...       7.000000       2.000000       3.000000   

          OTROMUERTO     OTROHERIDO     TOTMUERTOS     TOTHERIDOS  \
count  180219.000000  180219.000000  180219.000000  180219.000000   
mean        0.000022       0.000283       0.006331       0.188970   
std         0.004711       0.018695       0.087753       0.604156   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       2.000000       6.000000      32.000000   

               CLASE       LONGITUD        LATITUD  
count  180219.000000  180219.000000  180219.000000  
mean        2.851137    -101.874718      24.044809  
std         0.372082       4.859086       3.506709  
min         1.000000    -116.709410      14.857960  
25%         3.000000    -103.391610      20.657340  
50%         3.000000    -100.375640      25.631370  
75%         3.000000    -100.219350      25.747780  
max         3.000000     -86.746330      32.670720  

[8 rows x 45 columns]

In [10]:
df.describe(include=['O'])

ID                      CALLE1                       CALLE2  \
count   180219                      180212                       175938   
unique   45850                       22976                        26904   
top        413  AVENIDA EUGENIO GARZA SADA  CALZADA FRANCISCO I. MADERO   
freq        38                        1907                         1141   

       CARRETERA  
count       1420  
unique        25  
top               
freq        1382

### Resumen

**180,219 datos de 49 columnas**
Año 2019 únicamente

**Datos con nulos**:
- CALLE1: 7
- CALLE2: 4,281
- CARRETERA: 178,799

# Datos de CDMX

In [11]:
df2.head() 

id no_folio fecha_evento  ano_evento    mes   hora  condicion  \
0   0   906379   2019-01-01        2019  Enero  10:13  LESIONADO   
1   1   906166   2019-01-01        2019  Enero   7:39  LESIONADO   
2   2   906349   2019-01-01        2019  Enero   9:58  LESIONADO   
3   3   906469   2019-01-01        2019  Enero  11:35  LESIONADO   
4   4   907106   2019-01-01        2019  Enero  19:26  LESIONADO   

  tipo_de_evento  coordenada_x  coordenada_y  ...  identidad  \
0    ATROPELLADO    -99.143964     19.307892  ...     PEATON   
1         CHOQUE    -99.136425     19.470067  ...  CONDUCTOR   
2         CHOQUE    -99.069299     19.411029  ...  CONDUCTOR   
3         CHOQUE    -99.124262     19.470372  ...  CONDUCTOR   
4         CHOQUE    -98.996834     19.348251  ...  CONDUCTOR   

  unidad_medica_de_apoyo lugar_del_deceso trasladado_s_lesionado  \
0                   ERUM              NaN                    1.0   
1                   ERUM              NaN                    0.0   
2              CRUZ ROJA              NaN                    0.0   
3              CRUZ ROJA              NaN                    1.0   
4                   ERUM              NaN                    0.0   

                 hospital     dia hora2 prioridad observaciones  \
0  PEDIATRICO DE COYOACAN  martes  10.0     MEDIA           NaN   
1                     NaN  martes   7.0      BAJA           NaN   
2                     NaN  martes   9.0      BAJA         POSTE   
3      ANGELES LINDAVISTA  martes  11.0     MEDIA  SEÑALAMIENTO   
4                     NaN  martes  19.0      BAJA           NaN   

             coordenada  
0  19.307892,-99.143964  
1  19.470067,-99.136425  
2  19.411029,-99.069299  
3  19.470372,-99.124262  
4  19.348251,-98.996834  

[5 rows x 62 columns]

In [12]:
df2.tail() 

id no_folio fecha_evento  ano_evento    mes            hora  \
30753  30753   567604   2018-06-29        2018  Junio  08:12:00 p. m.   
30754  30754   567773   2018-06-29        2018  Junio  09:53:00 p. m.   
30755  30755   566355   2018-06-29        2018  Junio  04:38:00 a. m.   
30756  30756   567760   2018-06-29        2018  Junio  09:46:00 p. m.   
30757  30757   567893   2018-06-29        2018  Junio  11:29:00 p. m.   

       condicion tipo_de_evento  coordenada_x  coordenada_y  ...  \
30753  LESIONADO         CHOQUE    -99.216789     19.316302  ...   
30754  LESIONADO         CHOQUE    -99.176464     19.392405  ...   
30755  LESIONADO      DERRAPADO    -99.144095     19.342768  ...   
30756  LESIONADO      DERRAPADO    -99.133843     19.420006  ...   
30757  LESIONADO      DERRAPADO    -99.097561     19.479041  ...   

          identidad unidad_medica_de_apoyo lugar_del_deceso  \
30753  MOTOCICLISTA                PM/ERUM              NaN   
30754      CICLISTA                PM/ERUM              NaN   
30755  MOTOCICLISTA                   ERUM              NaN   
30756  MOTOCICLISTA                PM/ERUM              NaN   
30757  MOTOCICLISTA                   ERUM              NaN   

      trasladado_s_lesionado hospital      dia hora2 prioridad observaciones  \
30753                    1.0       SD  viernes  20.0     MEDIA           NaN   
30754                    0.0      NaN  viernes  21.0      BAJA           NaN   
30755                    0.0      NaN  viernes   4.0      BAJA           NaN   
30756                    0.0      NaN  viernes  21.0      BAJA           NaN   
30757                    0.0      NaN  viernes  23.0      BAJA           NaN   

                   coordenada  
30753    19.316302,-99.216789  
30754  19.3924048,-99.1764639  
30755    19.342768,-99.144095  
30756  19.4200064,-99.1338433  
30757  19.4790407,-99.0975605  

[5 rows x 62 columns]

In [13]:
df2.shape 

(30758, 62)

In [14]:
df2.columns

Index(['id', 'no_folio', 'fecha_evento', 'ano_evento', 'mes', 'hora',
       'condicion', 'tipo_de_evento', 'coordenada_x', 'coordenada_y',
       'punto_1', 'punto_2', 'colonia', 'alcaldia', 'tipo_de_interseccion',
       'tipo_de_vehiculo_1', 'tipo_de_vehiculo_2', 'tipo_de_vehiculo_3',
       'tipo_de_vehiculo_4', 'marca_de_vehiculo_1', 'marca_de_vehiculo_2',
       'marca_de_vehiculo_3', 'marca_de_vehiculo_4', 'lesiones', 'edad_occiso',
       'edad_lesionado', 'se_desconoce_occiso', 'se_desconoce_lesionado',
       'rango_de_edades_menores_de_1_ano_occiso',
       'rango_de_edades_menores_de_1_ano_lesionado', 'rango_edades_1_4_occiso',
       'rango_edades_1_4_lesionado', 'rango_edades_5_14_occiso',
       'rango_edades_5_14_lesionado', 'rango_edades_15_24_occiso',
       'rango_edades_15_24_lesionado', 'rango_edades_25_34_occiso',
       'rango_edades_25_34_lesionado', 'rango_edades_35_44_occiso',
       'rango_edades_35_44_lesionado', 'rango_edades_45_64_occiso',
       'rango_ed

In [15]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30758 entries, 0 to 30757
Data columns (total 62 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   id                                          30758 non-null  int64  
 1   no_folio                                    30758 non-null  object 
 2   fecha_evento                                30758 non-null  object 
 3   ano_evento                                  30758 non-null  int64  
 4   mes                                         30758 non-null  object 
 5   hora                                        30758 non-null  object 
 6   condicion                                   30758 non-null  object 
 7   tipo_de_evento                              30758 non-null  object 
 8   coordenada_x                                30758 non-null  float64
 9   coordenada_y                                30758 non-null  float64
 10  punto_1   

In [16]:
df2.describe()

id    ano_evento  coordenada_x  coordenada_y  \
count  30758.000000  30758.000000  30758.000000  30758.000000   
mean   15378.500000   2018.446388    -99.140562     19.392481   
std     8879.214126      0.497126      0.069885      0.067229   
min        0.000000   2018.000000    -99.999630     18.893255   
25%     7689.250000   2018.000000    -99.175600     19.346810   
50%    15378.500000   2018.000000    -99.141451     19.398742   
75%    23067.750000   2019.000000    -99.105318     19.440424   
max    30757.000000   2019.000000    -98.153068     19.963880   

       se_desconoce_occiso  se_desconoce_lesionado  \
count         30758.000000            30758.000000   
mean              0.005787                0.142044   
std               0.107430                0.436582   
min               0.000000                0.000000   
25%               0.000000                0.000000   
50%               0.000000                0.000000   
75%               0.000000                0.000000   
max              10.000000               16.000000   

       rango_de_edades_menores_de_1_ano_occiso  \
count                                  30758.0   
mean                                       0.0   
std                                        0.0   
min                                        0.0   
25%                                        0.0   
50%                                        0.0   
75%                                        0.0   
max                                        0.0   

       rango_de_edades_menores_de_1_ano_lesionado  rango_edades_1_4_occiso  \
count                                30758.000000             30758.000000   
mean                                     0.001691                 0.000065   
std                                      0.041867                 0.008064   
min                                      0.000000                 0.000000   
25%                                      0.000000                 0.000000   
50%                                      0.000000                 0.000000   
75%                                      0.000000                 0.000000   
max                                      2.000000                 1.000000   

       rango_edades_1_4_lesionado  ...  total_occisos  occisos_femeninos  \
count                30758.000000  ...   30758.000000       30758.000000   
mean                     0.011639  ...       0.025717           0.005137   
std                      0.112870  ...       0.182345           0.074167   
min                      0.000000  ...       0.000000           0.000000   
25%                      0.000000  ...       0.000000           0.000000   
50%                      0.000000  ...       0.000000           0.000000   
75%                      0.000000  ...       0.000000           0.000000   
max                      3.000000  ...      10.000000           3.000000   

       occisos_masculinos  occiso_se_desconoce  total_lesionados  \
count        30758.000000         30758.000000      30758.000000   
mean             0.019605             0.000975          1.164152   
std              0.144158             0.070292          0.641677   
min              0.000000             0.000000          0.000000   
25%              0.000000             0.000000          1.000000   
50%              0.000000             0.000000          1.000000   
75%              0.000000             0.000000          1.000000   
max              3.000000            10.000000         20.000000   

       lesionados_femeninos  lesionados_masculinos  lesionado_se_desconoce  \
count          30758.000000           30758.000000            30758.000000   
mean               0.362052               0.781943                0.020157   
std                0.613230               0.578705                0.217089   
min                0.000000               0.000000                0.000000   
25%                0.000000               0.000000                0.000000   
50%                0.000

In [17]:
df2.describe(include=['O'])

no_folio fecha_evento        mes            hora  condicion  \
count     30758        30758      30758           30758      30758   
unique    23191          730         12            2879          2   
top          SD   2018-08-18  Diciembre  08:00:00 a. m.  LESIONADO   
freq       7275           93       3510              34      30028   

       tipo_de_evento            punto_1            punto_2 colonia  \
count           30758              30758              30758   30758   
unique              6               4356               6657    2253   
top            CHOQUE  ANILLO PERIFERICO  ANILLO PERIFERICO  CENTRO   
freq            15962               1215                392    1035   

          alcaldia  ... edad_occiso edad_lesionado     identidad  \
count        30758  ...         808          30159         30758   
unique          16  ...         104           2619            22   
top     CUAUHTEMOC  ...          SD             SD  MOTOCICLISTA   
freq          5001  ...         128           3174         11091   

       unidad_medica_de_apoyo           lugar_del_deceso hospital      dia  \
count                   30758                        730     8985    30758   
unique                    456                          3      542        7   
top                      ERUM  EN EL LUGAR DE LOS HECHOS       SD  viernes   
freq                    10979                        718     5191     4788   

       prioridad observaciones            coordenada  
count      30758          2840                 30758  
unique         3           350                 30642  
top         BAJA            SD  19.422542,-99.122396  
freq       20931           837                     2  

[4 rows x 30 columns]

### Resumen

**30,758 datos de 62 columnas**
Años 2018 y 2019

**Datos con nulos**:
- tipo_vehiculo_2: 14,777
- tipo_vehiculo_3: 30,096
- tipo_vehiculo_4: 30,678
- marca_vehiculo_1: 3
- marca_vehiculo_2: 29,228
- marca_vehiculo_3: 30,105
- marca_vehiculo_4: 30,679
- edad_occiso: 29,950
- edad_lesionado: 599
- lugar_del_deceso: 30,028
- hospital: 21773
- hora2: 8
- observaciones: 27,918

---

# Data Cleaning

## Dataset nacional

In [18]:
df[['CALLE1', 'CALLE2', 'CARRETERA', 'LATITUD', 'LONGITUD']].head()

CALLE1                        CALLE2  \
0  BOULEVARD LICENCIADO ADOLFO RUÍZ CORTINES           CALLE PASEO DEL RÍO   
1            AVENIDA PASEO DE LA SOLIDARIDAD  AVENIDA AGUASCALIENTES NORTE   
2                        AVENIDA UNIVERSIDAD                  CALLE ABEDUL   
3                     AVENIDA MAHATMA GANDHI        CALLE ABRAHAM GONZÁLEZ   
4             AVENIDA GLORIETA BENITO JUÁREZ       CALLE IGNACIO T. CHÁVEZ   

  CARRETERA   LATITUD   LONGITUD  
0       NaN  21.87119 -102.31962  
1       NaN  21.91522 -102.28652  
2       NaN  21.90032 -102.30621  
3       NaN  21.84853 -102.29386  
4       NaN  21.87392 -102.30055

In [19]:
df = df.drop('CARRETERA', axis=1)

In [20]:
calle1_lost = df[df['CALLE1'].isnull()][['CALLE1', 'CALLE2', 'LONGITUD', 'LATITUD']]
idx_null = df[df['CALLE1'].isnull()].index

print("Registros sin valores en CALLE1")
calle1_lost

Registros sin valores en CALLE1


CALLE1 CALLE2    LONGITUD    LATITUD
17990    NaN    NaN -103.591160  19.062442
18001    NaN    NaN -103.526209  19.194464
18357    NaN    NaN -103.644521  19.152636
18361    NaN    NaN -103.726436  19.145940
18479    NaN    NaN -103.723074  19.193696
18842    NaN    NaN -103.599862  19.080913
18924    NaN    NaN -103.723161  19.192421

In [21]:
from geopy.geocoders import Nominatim
from time import sleep
import re

calle1_lost[['LATITUD', 'LONGITUD']]

geolocator = Nominatim(user_agent="geoapi")

regex = r"([^,]+), [^,]+, Municipio de ([^,]+), ([^,]+),"
locations = []

for index, row in calle1_lost.iterrows():
    lat = row['LATITUD']
    lon = row['LONGITUD']
    
    try:
        location = geolocator.reverse((lat, lon), language="es")
        direccion = location.address if location else "No se encontró información"
        # print(direccion + '\n')
        match = re.search(regex, direccion)
        if match:
            street = match.group(1).strip()
            locations.append(street)
        else:
            locations.append("Calle sin nombre")
        
        sleep(1)
    except Exception as e:
        print(f"Error con las coordenadas ({lat}, {lon}): {e}")
        
# Asignar los valores de 'locations' a índices en la columna 'CALLE1' sin valores
df.loc[idx_null, 'CALLE1'] = locations

In [22]:
df['CALLE1'].isnull().sum()

0

In [23]:
df.duplicated().sum()

0

De momento la columna de `CALLE2` no se va a eliminar en caso de necesitarla más adelante, pero de momento sólo se va a trabajar con la columna de `CALLE1`.

In [24]:
# df = df.drop('CALLE2', axis=1)

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180219 entries, 0 to 180218
Data columns (total 48 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   ID          180219 non-null  object 
 1   EDO         180219 non-null  int64  
 2   MES         180219 non-null  int64  
 3   ANIO        180219 non-null  int64  
 4   MPIO        180219 non-null  int64  
 5   HORA        180219 non-null  int64  
 6   MINUTOS     180219 non-null  int64  
 7   DIA         180219 non-null  int64  
 8   DIASEMANA   180219 non-null  int64  
 9   URBANA      180219 non-null  int64  
 10  SUBURBANA   180219 non-null  int64  
 11  TIPACCID    180219 non-null  int64  
 12  AUTOMOVIL   180219 non-null  int64  
 13  CAMPASAJ    180219 non-null  int64  
 14  MICROBUS    180219 non-null  int64  
 15  PASCAMION   180219 non-null  int64  
 16  OMNIBUS     180219 non-null  int64  
 17  TRANVIA     180219 non-null  int64  
 18  CAMIONETA   180219 non-null  int64  
 19  CA

## Dataset CDMX

In [26]:
null_cols = df2.columns[df2.isnull().any()]
null_cols

Index(['tipo_de_vehiculo_2', 'tipo_de_vehiculo_3', 'tipo_de_vehiculo_4',
       'marca_de_vehiculo_1', 'marca_de_vehiculo_2', 'marca_de_vehiculo_3',
       'marca_de_vehiculo_4', 'edad_occiso', 'edad_lesionado',
       'lugar_del_deceso', 'hospital', 'hora2', 'observaciones'],
      dtype='object')

In [27]:
df2[null_cols].head()

tipo_de_vehiculo_2 tipo_de_vehiculo_3 tipo_de_vehiculo_4  \
0                NaN                NaN                NaN   
1          AUTOMOVIL                NaN                NaN   
2        OBJETO FIJO                NaN                NaN   
3        OBJETO FIJO                NaN                NaN   
4          AUTOMOVIL                NaN                NaN   

  marca_de_vehiculo_1 marca_de_vehiculo_2 marca_de_vehiculo_3  \
0               DODGE                 NaN                 NaN   
1               DERBY               VERSA                 NaN   
2               URVAN                 NaN                 NaN   
3             JOURNEY                 NaN                 NaN   
4               CHEVY               SPARK                 NaN   

  marca_de_vehiculo_4 edad_occiso edad_lesionado lugar_del_deceso  \
0                 NaN         NaN              9              NaN   
1                 NaN         NaN          39 25              NaN   
2                 NaN         NaN             43              NaN   
3                 NaN         NaN             SD              NaN   
4                 NaN         NaN             28              NaN   

                 hospital  hora2 observaciones  
0  PEDIATRICO DE COYOACAN   10.0           NaN  
1                     NaN    7.0           NaN  
2                     NaN    9.0         POSTE  
3      ANGELES LINDAVISTA   11.0  SEÑALAMIENTO  
4                     NaN   19.0           NaN

In [28]:
df2[null_cols].describe()

hora2
count  30750.000000
mean      12.572780
std        6.239031
min        0.000000
25%        8.000000
50%       13.000000
75%       18.000000
max       23.000000

In [29]:
df2[null_cols].describe(include=['O'])

tipo_de_vehiculo_2 tipo_de_vehiculo_3 tipo_de_vehiculo_4  \
count               15981                661                 80   
unique                 15                 11                  9   
top             AUTOMOVIL          AUTOMOVIL          AUTOMOVIL   
freq                 5835                432                 55   

       marca_de_vehiculo_1 marca_de_vehiculo_2 marca_de_vehiculo_3  \
count                30755               15302                 653   
unique                 870                 594                 151   
top                     SD                  SD                  SD   
freq                 10494                5893                 114   

       marca_de_vehiculo_4 edad_occiso edad_lesionado  \
count                   79         808          30159   
unique                  46         104           2619   
top                     SD          SD             SD   
freq                     9         128           3174   

                 lugar_del_deceso hospital observaciones  
count                         730     8985          2840  
unique                          3      542           350  
top     EN EL LUGAR DE LOS HECHOS       SD            SD  
freq                          718     5191           837

In [30]:
df2 = df2.drop(
    [
        'tipo_de_vehiculo_2',
        'tipo_de_vehiculo_3', 
        'tipo_de_vehiculo_4',
        'marca_de_vehiculo_2', 
        'marca_de_vehiculo_3', 
        'marca_de_vehiculo_4',
        'lugar_del_deceso',
        'hora2'
    ],
    axis=1
)

In [31]:
injured_cols = [
    'edad_lesionado', 'se_desconoce_lesionado',
    'rango_de_edades_menores_de_1_ano_lesionado', 'rango_edades_1_4_lesionado', 
    'rango_edades_5_14_lesionado', 'rango_edades_15_24_lesionado', 
    'rango_edades_25_34_lesionado', 'rango_edades_35_44_lesionado', 
    'rango_edades_45_64_lesionado', 'rango_edades_65_lesionado',  'total_lesionados',
    'lesionados_femeninos', 'lesionados_masculinos', 'lesionado_se_desconoce'
]

deceased_cols = [
    'edad_occiso', 'se_desconoce_occiso', 'rango_de_edades_menores_de_1_ano_occiso',
    'rango_edades_1_4_occiso', 'rango_edades_5_14_occiso', 'rango_edades_15_24_occiso', 
    'rango_edades_25_34_occiso', 'rango_edades_35_44_occiso',
    'rango_edades_45_64_occiso', 'rango_edades_65_occiso', 'total_occisos', 
    'occisos_femeninos', 'occisos_masculinos', 'occiso_se_desconoce',
]

In [32]:
df2[injured_cols].head()

edad_lesionado  se_desconoce_lesionado  \
0              9                       0   
1          39 25                       0   
2             43                       0   
3             SD                       1   
4             28                       0   

   rango_de_edades_menores_de_1_ano_lesionado  rango_edades_1_4_lesionado  \
0                                           0                           0   
1                                           0                           0   
2                                           0                           0   
3                                           0                           0   
4                                           0                           0   

   rango_edades_5_14_lesionado  rango_edades_15_24_lesionado  \
0                            1                             0   
1                            0                             0   
2                            0                             0   
3                            0                             0   
4                            0                             0   

   rango_edades_25_34_lesionado  rango_edades_35_44_lesionado  \
0                             0                             0   
1                             1                             1   
2                             0                             1   
3                             0                             0   
4                             1                             0   

   rango_edades_45_64_lesionado  rango_edades_65_lesionado  total_lesionados  \
0                             0                          0                 1   
1                             0                          0                 2   
2                             0                          0                 1   
3                             0                          0                 1   
4                             0                          0                 1   

   lesionados_femeninos  lesionados_masculinos  lesionado_se_desconoce  
0                     0                      1                       0  
1                     0                      2                       0  
2                     0                      1                       0  
3                     1                      0                       0  
4                     0                      1                       0

In [33]:
df2[deceased_cols].head()

edad_occiso  se_desconoce_occiso  rango_de_edades_menores_de_1_ano_occiso  \
0         NaN                    0                                        0   
1         NaN                    0                                        0   
2         NaN                    0                                        0   
3         NaN                    0                                        0   
4         NaN                    0                                        0   

   rango_edades_1_4_occiso  rango_edades_5_14_occiso  \
0                        0                         0   
1                        0                         0   
2                        0                         0   
3                        0                         0   
4                        0                         0   

   rango_edades_15_24_occiso  rango_edades_25_34_occiso  \
0                          0                          0   
1                          0                          0   
2                          0                          0   
3                          0                          0   
4                          0                          0   

   rango_edades_35_44_occiso  rango_edades_45_64_occiso  \
0                          0                          0   
1                          0                          0   
2                          0                          0   
3                          0                          0   
4                          0                          0   

   rango_edades_65_occiso  total_occisos  occisos_femeninos  \
0                       0              0                  0   
1                       0              0                  0   
2                       0              0                  0   
3                       0              0                  0   
4                       0              0                  0   

   occisos_masculinos  occiso_se_desconoce  
0                   0                    0  
1                   0                    0  
2                   0                    0  
3                   0                    0  
4                   0                    0

Para encontrar la edad de los lesionados/occisos se puede hacer a partir de los rangos de edad y del total de lesionados/occisos.
En el caso de los occisos, la mayoría de los datos faltantes deben ser 0, ya que en esos casos el total de occisos es de 0, por eso esos valores son nulos.
Para ambos casos, es objeto ya que en algunos casos hay más de una edad.
El dato "SD" en algunas columnas significa **Se desconoce**.

In [34]:
df2.loc[df2["total_occisos"] == 0, "edad_occiso"] = 0

In [35]:
df2.loc[df2["total_lesionados"] == 0, "edad_lesionado"] = 0

In [36]:
edad_lesionado_lost = df2[df2['edad_lesionado'].isnull()][[
    'identidad', 'lesionado_se_desconoce', 'se_desconoce_lesionado', 
    'rango_de_edades_menores_de_1_ano_lesionado', 'rango_edades_1_4_lesionado', 
    'rango_edades_5_14_lesionado', 'rango_edades_15_24_lesionado', 
    'rango_edades_25_34_lesionado', 'rango_edades_35_44_lesionado', 
    'rango_edades_45_64_lesionado', 'rango_edades_65_lesionado',
    'total_lesionados', 'total_occisos', 'lesionados_femeninos', 'lesionados_masculinos'
]]
idx_lesioando_null = df2[df2['edad_lesionado'].isnull()].index

print("Registros sin valores en edad_lesionado")
edad_lesionado_lost

Registros sin valores en edad_lesionado


identidad  lesionado_se_desconoce  se_desconoce_lesionado  \
108        CICLISTA                       0                       0   
10832  MOTOCICLISTA                       0                       0   
11288      PASAJERO                       0                       1   
24794  MOTOCICLISTA                       0                       1   
24796     CONDUCTOR                       0                       1   

       rango_de_edades_menores_de_1_ano_lesionado  rango_edades_1_4_lesionado  \
108                                             0                           0   
10832                                           0                           0   
11288                                           0                           0   
24794                                           0                           0   
24796                                           0                           0   

       rango_edades_5_14_lesionado  rango_edades_15_24_lesionado  \
108                              0                             0   
10832                            0                             0   
11288                            0                             0   
24794                            0                             0   
24796                            0                             0   

       rango_edades_25_34_lesionado  rango_edades_35_44_lesionado  \
108                               0                             1   
10832                             1                             0   
11288                             0                             0   
24794                             0                             0   
24796                             0                             0   

       rango_edades_45_64_lesionado  rango_edades_65_lesionado  \
108                               0                          0   
10832                             0                          0   
11288                             0                          0   
24794                             0                          0   
24796                             0                          0   

       total_lesionados  total_occisos  lesionados_femeninos  \
108                   1              0                     0   
10832                 1              0                     0   
11288                 1              1                     0   
24794                 1              0                     0   
24796                 1              0                     0   

       lesionados_masculinos  
108                        1  
10832                      1  
11288                      1  
24794                      1  
24796                      1

In [37]:
df2.loc[df2["se_desconoce_lesionado"] == 0, "edad_lesionado"] = 0

In [38]:
df2.drop(index=[11288, 24794, 24796], inplace=True)

In [39]:
marca_vehiculo_lost = df2[df2['marca_de_vehiculo_1'].isnull()][[
    'identidad', 'tipo_de_vehiculo_1', 'tipo_de_evento', 'observaciones'
]]
idx_marca_null = df2[df2['marca_de_vehiculo_1'].isnull()].index

print("Registros sin valores en marca_de_vehiculo_1")
marca_vehiculo_lost

Registros sin valores en marca_de_vehiculo_1


identidad tipo_de_vehiculo_1 tipo_de_evento observaciones
2321   MOTOCICLISTA          AUTOMOVIL         CHOQUE     MOTO TAXI
7105         PEATON                 SD    ATROPELLADO           NaN
13291  MOTOCICLISTA        MOTOCICLETA         CHOQUE           NaN

In [40]:
df2['marca_de_vehiculo_1'].value_counts().head()

marca_de_vehiculo_1
SD         10492
ITALIKA     2374
YAMAHA      1300
HONDA       1233
VENTO        819
Name: count, dtype: int64

In [41]:
df2.loc[idx_marca_null, 'marca_de_vehiculo_1'] = ['ITALIKA', 'HONDA', 'ITALIKA']

In [42]:
df2['unidad_medica_de_apoyo'].unique()

array(['ERUM', 'CRUZ ROJA', 'CCO/CRUZ ROJA', 'CCO/ERUM', 'CCO/SPM',
       'CCO/MEDICA LIFE', 'CCO/PC', 'CCO/UNIDAD MEDICA',
       'CCO/AMT DE URGENCIAS BASICA', 'CCO/SD', 'CCO/CRUM',
       'CCO/1ER CONTACTO', 'CCO/FUEGO NUEVO', 'CCO/URGENCIAS BASICAS',
       'CCO/RESCATE', 'CCO/MDA', 'CCO/ERUM AIF', 'CCO/VULCANOS',
       'CCO/SAMU', 'CCO/NEZAHUALCOYOTL', 'CCO/AVDI', 'SD', 'MEDICAL',
       'CEBES MEDIC', 'PC', 'VULCANOS', 'PARAMEDIC', 'MADECAR',
       'PARTICULAR', 'MEDIC', 'SAMU', 'CRUM', 'AIEP', 'ALFAMEDIC',
       'CCO/USI', 'URGENCIAS BASICAS', 'SEM', 'SOPORTE', 'SUMA',
       'VITALET', 'MEDICAR', 'TRITON', 'CIEL MEDIC', 'TRIMEDIC',
       'RESCATE XOCHIMILCO', 'MEDICA ANGELES', 'AVEFREE', 'UMET',
       'PRIMER CONTACTO', 'SOPORTE VITAL', 'PBI', 'SEAS', 'SUPORT',
       'MEDIC WAN', 'UNIPOL', 'VECTOR', 'SIMA', 'BENGALA', 'AMETRE',
       'CONTACTO', 'UMEC', 'URGEMEDIC', 'SIBA', 'SEPROMEDIC', 'EUROJET',
       'CUIDADOS INTENSIVOS', 'GALILEOS', 'CICROMEDIC', 'ARE',
       'F

In [43]:
df2['hospital'].head()

0    PEDIATRICO DE COYOACAN
1                       NaN
2                       NaN
3        ANGELES LINDAVISTA
4                       NaN
Name: hospital, dtype: object

In [44]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30755 entries, 0 to 30757
Data columns (total 54 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   id                                          30755 non-null  int64  
 1   no_folio                                    30755 non-null  object 
 2   fecha_evento                                30755 non-null  object 
 3   ano_evento                                  30755 non-null  int64  
 4   mes                                         30755 non-null  object 
 5   hora                                        30755 non-null  object 
 6   condicion                                   30755 non-null  object 
 7   tipo_de_evento                              30755 non-null  object 
 8   coordenada_x                                30755 non-null  float64
 9   coordenada_y                                30755 non-null  float64
 10  punto_1        

Debido a que las variables de `hospital` y `observaciones` tienen bastantes nulos y no hay manera rápida de imputar datos, se quedará con esos datos nulos, ya que ambas columnas podrían ser de utilidad.

In [45]:
object_cols = df2.select_dtypes(include=[object]).columns
object_cols

Index(['no_folio', 'fecha_evento', 'mes', 'hora', 'condicion',
       'tipo_de_evento', 'punto_1', 'punto_2', 'colonia', 'alcaldia',
       'tipo_de_interseccion', 'tipo_de_vehiculo_1', 'marca_de_vehiculo_1',
       'lesiones', 'edad_occiso', 'edad_lesionado', 'identidad',
       'unidad_medica_de_apoyo', 'hospital', 'dia', 'prioridad',
       'observaciones', 'coordenada'],
      dtype='object')

In [46]:
df2[object_cols].head()

no_folio fecha_evento    mes   hora  condicion tipo_de_evento  \
0   906379   2019-01-01  Enero  10:13  LESIONADO    ATROPELLADO   
1   906166   2019-01-01  Enero   7:39  LESIONADO         CHOQUE   
2   906349   2019-01-01  Enero   9:58  LESIONADO         CHOQUE   
3   906469   2019-01-01  Enero  11:35  LESIONADO         CHOQUE   
4   907106   2019-01-01  Enero  19:26  LESIONADO         CHOQUE   

                    punto_1                punto_2             colonia  \
0                   HIDALGO        PRIV DE HIDALGO    STA URSULA COAPA   
1               EJE CENTRAL         AV INSURGENTES             VALLEJO   
2                 EJE 5 PTE                 AV NTE  AGRICOLA PANTITLAN   
3  AV BASILIO ROMO ANGUIANO  CALZ DE LOS MISTERIOS             TEPEYAC   
4          15 DE SEPTIEMBRE              EJE 6 SUR   LOMAS DE ZARAGOZA   

           alcaldia  ...                       lesiones edad_occiso  \
0          COYOACAN  ...             LUMBALGIA POSTURAL           0   
1  GUSTAVO A MADERO  ...  POLICONTUNDIDO POLICONTUNDIDO           0   
2         IZTACALCO  ...                 POLICONTUNDIDO           0   
3  GUSTAVO A MADERO  ...                 POLICONTUNDIDO           0   
4        IZTAPALAPA  ...                CRISIS NERVIOSA           0   

  edad_lesionado  identidad unidad_medica_de_apoyo                hospital  \
0              0     PEATON                   ERUM  PEDIATRICO DE COYOACAN   
1              0  CONDUCTOR                   ERUM                     NaN   
2              0  CONDUCTOR              CRUZ ROJA                     NaN   
3             SD  CONDUCTOR              CRUZ ROJA      ANGELES LINDAVISTA   
4              0  CONDUCTOR                   ERUM                     NaN   

      dia prioridad observaciones            coordenada  
0  martes     MEDIA           NaN  19.307892,-99.143964  
1  martes      BAJA           NaN  19.470067,-99.136425  
2  martes      BAJA         POSTE  19.411029,-99.069299  
3  martes     MEDIA  SEÑALAMIENTO  19.470372,-99.124262  
4  martes      BAJA           NaN  19.348251,-98.996834  

[5 rows x 23 columns]

In [47]:
df2.duplicated().sum()

0

In [48]:
df = df.drop('ID', axis=1)
df = df.drop('CALLE2', axis=1)

In [49]:
df.to_csv('../data/processed/cleaned_nacional.csv', index=False)
df2.to_csv('../data/processed/cleaned_cdmx.csv', index=False)